mne/morph.py

```diff
@@ -0,0 +1,998 @@
+                                  starting_affine=c_of_mass.affine)
+
+    # rigid body transform (translation + rotation)
+    rigid = affreg.optimize(mri_to, mri_from,
+                            transforms.RigidTransform3D(), None,
+                            mri_to_grid2world, mri_from_grid2world,
+                            starting_affine=translation.affine)
+
+    # affine transform (translation + rotation + scaling)
+    affine = affreg.optimize(mri_to, mri_from,
+                             transforms.AffineTransform3D(), None,
+                             mri_to_grid2world, mri_from_grid2world,
+                             starting_affine=rigid.affine)
+
+    # apply affine transformation
+    mri_from_affine = affine.transform(mri_from)
+
+    # set up Symmetric Diffeomorphic Registration (metric, iterations)
+    sdr = imwarp.SymmetricDiffeomorphicRegistration(
+        metrics.CCMetric(3), list(niter_sdr))
+
+    # compute mapping
+    mapping = sdr.optimize(mri_to, mri_from_affine)
+
+    morph['shape'] = tuple(mapping.domain_shape.astype('float'))
+    morph['zooms'] = grid_spacing
+    morph['affine'] = mri_to_grid2world
+    morph['AffineMap'] = affine.__dict__
+    morph['DiffeomorphicMap'] = mapping.__dict__
+
+    logger.info('done.')
+
+    return morph
+
+
+###############################################################################
+# Morph for SourceEstimate | VectorSourceEstimate
+
+@verbose
+def _morph_buffer(data, idx_use, e, smooth, n_vertices, nearest, maps,
+                  warn=True, verbose=None):
+    """Morph data from one subject's source space to another.
+
+    Parameters
+    ----------
+    data : array, or csr sparse matrix
+        A n_vertices [x 3] x n_times (or other dimension) dataset to morph.
+    idx_use : array of int
+        Vertices from the original subject's data.
+    e : sparse matrix
+        The mesh edges of the "from" subject.
+    smooth : int
+        Number of smoothing iterations to perform. A hard limit of 100 is
+        also imposed.
+    n_vertices : int
+        Number of vertices.
+    nearest : array of int
+        Vertices on the destination surface to use.
+    maps : sparse matrix
+        Morph map from one subject to the other.
+    warn : bool
+        If True, warn if not all vertices were used.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see :func:`mne.verbose`
+        and :ref:`Logging documentation <tut_logging>` for more).
+
+    Returns
+    -------
+    data_morphed : array, or csr sparse matrix
+        The morphed data (same type as input).
+    """
+    # When operating on vector data, morph each dimension separately
+    if data.ndim == 3:
+        data_morphed = np.zeros((len(nearest), 3, data.shape[2]),
+                                dtype=data.dtype)
+        for dim in range(3):
+            data_morphed[:, dim, :] = _morph_buffer(
```

Comment:
is recursion really necessary here?

Reply:
no clue, just cp from ``source_estimate.py``